FirebaseAuth/Sources/Swift/Auth/AuthTokenResult.swift

```diff
@@ -121,56 +82,81 @@ private let kClaimsKey = "claims"
-    ) else {
-      return nil
-    }
-    guard let tokenPayloadDictionary = try? JSONSerialization.jsonObject(
-      with: decodedTokenPayloadData,
+  private class func getTokenPayloadDictionary(_ payloadData: Data) -> [String: Any]? {
+    return try? JSONSerialization.jsonObject(
+      with: payloadData,
       options: [.mutableContainers, .allowFragments]
-    ) as? [String: Any] else {
-      return nil
-    }
+    ) as? [String: Any]
+  }
+
+  private class func getJWT(_ payloadData: Data) -> JWT? {
     // These are dates since 00:00:00 January 1 1970, as described by the Terminology section in
     // the JWT spec. https://tools.ietf.org/html/rfc7519
     let decoder = JSONDecoder()
     decoder.dateDecodingStrategy = .secondsSince1970
     decoder.keyDecodingStrategy = .convertFromSnakeCase
-    guard let jwt = try? decoder.decode(JWT.self, from: decodedTokenPayloadData) else {
+    guard let jwt = try? decoder.decode(JWT.self, from: payloadData) else {
       return nil
     }
+    return jwt
+  }
 
-    let tokenResult = AuthTokenResult(token: token,
-                                      expirationDate: jwt.exp,
-                                      authDate: jwt.authTime,
-                                      issuedAtDate: jwt.iat,
-                                      signInProvider: jwt.firebase.signInProvider,
-                                      signInSecondFactor: jwt.firebase.signInSecondFactor,
-                                      claims: tokenPayloadDictionary)
-    return tokenResult
+  /** @fn tokenResultWithToken:
+       @brief Parse a token string to a structured token.
+       @param token The token string to parse.
+       @return A structured token result.
+   */
+  @objc open class func tokenResult(token: String) -> AuthTokenResult? {
+    guard let payloadData = getTokenPayloadData(token),
+          let claims = getTokenPayloadDictionary(payloadData),
+          let jwt = getJWT(payloadData) else {
+      return nil
+    }
+    return AuthTokenResult(token: token, jwt: jwt, claims: claims)
   }
 
-  init(token: String,
-       expirationDate: Date,
-       authDate: Date,
-       issuedAtDate: Date,
-       signInProvider: String,
-       signInSecondFactor: String?,
-       claims: [String: Any]) {
+  private init(token: String, jwt: JWT, claims: [String: Any]) {
     self.token = token
-    self.expirationDate = expirationDate
-    self.authDate = authDate
-    self.issuedAtDate = issuedAtDate
-    self.signInProvider = signInProvider
-    self.signInSecondFactor = signInSecondFactor ?? ""
+    expirationDate = jwt.exp
+    authDate = jwt.authTime
+    issuedAtDate = jwt.iat
+    signInProvider = jwt.firebase.signInProvider
+    signInSecondFactor = jwt.firebase.signInSecondFactor ?? ""
     self.claims = claims
   }
+
+  // MARK: Secure Coding
+
+  private static let kTokenKey = "token"
+
+  public static var supportsSecureCoding: Bool {
+    return true
+  }
```

Comment:
same here, potentially?
```suggestion
  static var supportsSecureCoding: Bool = true
```